### ISSUE TITLE
Optimization time does not change when JIT is used.

### ISSUE THREAD

**Issue by brunomorampc** [mechmotum/cyipopt#204]
I want to use JAX to solve the optimization problem you have in the documentation.
I noticed that the optimization takes much more time than if I solve the same problem using CasADi (https://web.casadi.org/) which also uses IPOPT ant it is used to compute the Jakobians and hessians with AD. Here is the profiler:

![image](https://github.com/mechmotum/cyipopt/assets/27070958/88ecd468-60c2-478e-a276-c855b424d7fe)

The dispatch.py is taking most of the time. As far as I understood, this is the one involved in the just-in-time compilation but I was expecting to be important only once the functions are called once.  After the first call, if I understood it correctly, since the input dimension of these functions do not change, JAX should use the cached functions. But this does not seem to occur now, since the optimization always takes the same amount of time and dispatch.py seems to be always the bottleneck.

**Comment by brocksam** [comment]
Can you please post a copy of the code snippet that you used to produce those profile results so that others are able to investigate? 

From your post above, `dispatch.py` isn't part of CyIpopt, I assume it's part of JAX. I'd like to know if there is a large overhead being incurred on each function call or if there is only a single 1.2s penalty for using JAX.

**Comment by brunomorampc** [comment]
Hi @brocksam thanks for answering
here's the code

```
import time
import numpy as np

from jax.config import config

# Enable 64 bit floating point precision
config.update("jax_enable_x64", True)

# We use the CPU instead of GPU und mute all warnings if no GPU/TPU is found.
config.update('jax_platform_name', 'cpu')

import jax.numpy as np
from jax import jit, grad, jacfwd, jacrev
from cyipopt import minimize_ipopt
import cProfile
import pstats
import io

profiler = cProfile.Profile()
def optimization():
    def objective(x):
        return x[0] * x[3] * np.sum(x[:3]) + sum(x)


    def eq_constraints(x):
        return np.sum(x ** 2) - 40


    def ineq_constrains(x):
        return np.prod(x[0:4]) - 25

    n_states = 50

    # jit the functions
    obj_jit = jit(objective)
    con_eq_jit = jit(eq_constraints)
    con_ineq_jit = jit(ineq_constrains)

    # build the derivatives and jit them
    obj_grad = jit(grad(obj_jit))  # objective gradient
    obj_hess = jit(jacrev(jacfwd(obj_jit)))  # objective hessian
    con_eq_jac = jit(jacfwd(con_eq_jit))  # jacobian
    con_ineq_jac = jit(jacfwd(con_ineq_jit))  # jacobian
    con_eq_hess = jacrev(jacfwd(con_eq_jit))  # hessian
    con_eq_hessvp = jit(lambda x, v: con_eq_hess(x) * v[0])  # hessian vector-product
    con_ineq_hess = jacrev(jacfwd(con_ineq_jit))  # hessian
    con_ineq_hessvp = jit(lambda x, v: con_ineq_hess(x) * v[0])  # hessian vector-product


    # constraints
    cons = [
        {'type': 'eq', 'fun': con_eq_jit, 'jac': con_eq_jac, 'hess': con_eq_hessvp},
        {'type': 'ineq', 'fun': con_ineq_jit, 'jac': con_ineq_jac, 'hess': con_ineq_hessvp}
    ]

    # starting point
    x0 = np.array([1.0 for _ in range(n_states)])

    # variable bounds: 1 <= x[i] <= 5
    bnds = [(0, 5) for _ in range(x0.size)]
    start = time.time()
    # executing the solver
    profiler.enable()
    res = minimize_ipopt(obj_jit, jac=obj_grad, hess=obj_hess, x0=x0, bounds=bnds,
                         constraints=cons, options={'disp': 5})
    profiler.disable()
    stats = pstats.Stats(profiler).sort_stats('cumtime')
    stats.dump_stats('profiler_jit.txt')
    print(f"Elapsed time {time.time() - start} seconds")

if __name__ == '__main__':
    optimization()
```

yes that function is part of JAX as far as I understood

**Comment by moorepants** [comment]
What if you evaluate the jitted functions once before passing to minimize_ipopt? I think you should only be profiling minimize_ipopt() also, as the other stuff is just setup.

**Comment by brunomorampc** [comment]
Hey @moorepants ,

thanks. I tried to evaluate the jitted functions first and this did indeed decrease the computational time almost 3 times. Still not as fast as casadi ( 1.4 seconds against 0.2 seconds with CasADi) but better than before :)
And yes I am profiling only the minimize_ipopt, since I am enabling and disabling the profiler around that call.

**Comment by moorepants** [comment]
We have done zero work to make anything with Jax fast in this package, there has only been the support for passing in Jax functions. Also the `minimize_ipopt()` function is also not built for speed, it is built for convenience and quick use. If you really care about speed then you should be using `Problem()` directly and carefully constructing you objective, constraints, gradients, jacobians, hessians etc. If you do that, you will likely be much faster than Casadi, as it has lots of overhead too. That said, we will take PRs that speed up `minimize_ipopt()`.

**Comment by brunomorampc** [comment]
Got it, thanks a lot for your work!

**Comment by moorepants** [comment]
I'm going to close this as resolved.

Good luck!

**Comment by brunomorampc** [comment]
@moorepants 
I actually have 2 other questions:
1) If I use the Problem class, can I still use jax to compute jacobian and hessians?
2) for what is scipy used in the minimize_ipopt function? If you use jax to compute hessisans and jacobians, shouln't you have all the information  needed to pass them directly to IPOPT? Why is scipy needed?
Thanks!

**Comment by moorepants** [comment]
> If I use the Problem class, can I still use jax to compute jacobian and hessians?

When using the Problem class you write your objective and constraints how every you like. The inputs and outputs of the functions have to be numpy arrays but you can do whatever you want inside those functions, including using Jax.

> for what is scipy used in the minimize_ipopt function? If you use jax to compute hessisans and jacobians, shouln't you have all the information needed to pass them directly to IPOPT? Why is scipy needed?

The 'minimize_ipopt` function was designed to be compatible with scipy's `minimize()` function and it uses scipy to achieve that.
